src/compose/app.ts

```diff
@@ -476,26 +570,13 @@ export class App {
 
 		for (const { current, target } of components) {
 			// If a current exists, we're either removing it or updating the configuration. In
-			// both cases, we must remove the component first, so we output those steps first.
+			// both cases, we must remove the component before creating it to avoid
+			// Engine conflicts. So we always emit a remove step first.
 			// If we do remove the component, we first need to remove any services which depend
-			// on the component
+			// on the component. The service removal steps are generated in this.generateStepsForService
+			// after their removal is calculated in this.compareServices.
 			if (current != null) {
-				// Find any services which are currently running which need to be killed when we
-				// recreate this component
-				const dependencies = _.filter(this.services, (s) =>
-					dependencyFn(current, s),
-				);
-				if (dependencies.length > 0) {
-					// We emit kill steps for these services, and wait to destroy the component in
-					// the next state application loop
-					// FIXME: We should add to the changingServices array, as we could emit several
-					// kill steps for a service
-					steps = steps.concat(
-						dependencies.flatMap((svc) =>
-							this.generateKillStep(svc, changingServices),
-						),
-					);
-				} else {
+				if (dependentServices.length === 0) {
```

Comment:
Not sure about this, wouldn't this prevent removal of a component even if there is another component that has dependent services? Example a volume A without dependencies would still wait for the removal of dependencies of a volume B. Is that the intention here?

Reply:
It would prevent it, but dependent services are destined for removal anyway so I don't think this is a big issue

Reply:
The problem is that the update would take longer, because a volume without dependencies can be immediately removed

Reply:
True. I modified the dependent service calculation to filter for dependent services of the current component (network/volume) only [here](https://github.com/balena-os/balena-supervisor/pull/2234/commits/f199319d0a2ecf3dcbfc59c82f022b662d784a71#diff-857bb737cdeb3c33e2caa352769d3f96e0e8a48da6fea254eb99bc627732cec1R568-R587)